graph/cytoscape/cytoscape.go

```diff
@@ -0,0 +1,200 @@
+// cytoscape provides conversion from graph/tree to cystoscape configuration json
+package cytoscape
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/swift-sunshine/swscore/graph/tree"
+	"github.com/swift-sunshine/swscore/log"
+)
+
+type NodeData struct {
+	Id            string `json:"id"`
+	Service       string `json:"service"`
+	Version       string `json:"version,omitempty"`
+	Text          string `json:"text"`
+	Parent        string `json:"parent,omitempty"`
+	LinkPromGraph string `json:"link_prom_graph,omitempty"`
+}
+
+type EdgeData struct {
+	Id     string `json:"id"`
+	Source string `json:"source"`
+	Target string `json:"target"`
+	Text   string `json:"text"`
+	Color  string `json:"color"`
+	Rpm    string `json:"req_per_min,omitempty"`
+	Rpm2xx string `json:"req_per_min_2XX,omitempty"`
+	Rpm3xx string `json:"req_per_min_3XX,omitempty"`
+	Rpm4xx string `json:"req_per_min_4XX,omitempty"`
+	Rpm5xx string `json:"req_per_min_5XX,omitempty"`
+}
+
+type NodeWrapper struct {
+	Data NodeData `json:"data"`
+}
+
+type EdgeWrapper struct {
+	Data EdgeData `json:"data"`
+}
+
+type Elements struct {
+	Nodes []*NodeWrapper `json:"nodes"`
+	Edges []*EdgeWrapper `json:"edges"`
+}
+
+type Config struct {
+	Elements Elements `json:"elements"`
+}
+
+// NewConfig currently ignores namespace arg
+func NewConfig(namespace string, sn *[]tree.ServiceNode) (result Config) {
+	nodes := []*NodeWrapper{}
+	edges := []*EdgeWrapper{}
+
+	var nodeIdSequence int
+	var edgeIdSequence int
+
+	for _, t := range *sn {
+		log.Debugf("Walk Tree Root %v", t.ID)
+
+		walk(&t, &nodes, &edges, "", &nodeIdSequence, &edgeIdSequence)
+	}
+
+	// Add composite nodes that group together different versions of the same service
+	addCompositeNodes(&nodes, &nodeIdSequence)
+
+	elements := Elements{nodes, edges}
+	result = Config{elements}
+	return result
+}
+
+func walk(sn *tree.ServiceNode, nodes *[]*NodeWrapper, edges *[]*EdgeWrapper, parentNodeId string, nodeIdSequence, edgeIdSequence *int) {
```

Comment:
Can we add some comments about strategy for building the graph ?
I.e. describing the raw input data, first to build the nodes, a second search for edges, or anything relevant to understand and maintain it ?
If not it's hard to understand if the algorithm is correct or not without running it with some examples.
Some description would help as well.

Reply:
+1 I will add more  high level description...